### ISSUE TITLE
error when trying to use msvc 2015: definition of dllimport static data member not allowed

### ISSUE THREAD

**Issue by doctor-yes** [randombit/botan#1677]
I process botan using this commands:
python ./configure.py --cc=msvc --cpu=x86 --minimized-build --single-amalgamation-file --amalgamation --enable-modules=rsa,auto_rng,sha1,emsa1,hmac,win32_stats 

and when I try to compile using msvc2015 (in Qt), I get these errors. How can I work around these? Thanks.

error: C2491: 'Botan::CPUID::g_processor_features': definition of dllimport static data member not allowed
error: C2491: 'Botan::CPUID::g_cache_line_size': definition of dllimport static data member not allowed
error: C2491: 'Botan::CPUID::g_endian_status': definition of dllimport static data member not allowed

**Comment by randombit** [comment]
> and when I try to compile using msvc2015 (in Qt)

What do you mean by "in Qt"? In Qt Creator? In any case I think this is the problem.  The makefile generated by configure sets the following preprocessor define `/DBOTAN_DLL=__declspec(dllexport)` to override the default setting (of `dllimport`, which is the value used in all situations other than the library itself being built, eg when one is compiling an application against the library). If this define was missing, the compiler would when compiling the library source see these vars as being `dllimport` static member variables, which indeed doesn't make any sense.

Probably all you need to do is configure Qt Creator to also redefine the `BOTAN_DLL` macro as above when building the library object.

(Also, do you really want a DLL here? Normally when one is using amalgamation the whole binary is included directly in the application.)

**Comment by Daedronus** [comment]
I think the bug here is that for amalgamation builds there should not be a dllexport/import

**Comment by doctor-yes** [comment]
Yes, by "Qt" I meant "Qt creator". So I confirm it is a problem of __declspec(dllexport) not being defined. Everything works if I make this change in the .h file.

#ifndef BOTAN_DLL
//    #define BOTAN_DLL __declspec(dllimport)
    #define BOTAN_DLL __declspec(dllexport)
#endif

You mentioned to configure Qt creator to also redefine the BOTAN_DLL macro... Can you give me some more detail where should I define what macro? Thanks a lot.
